### PR TITLE
Start of Categorical Scatterplot

### PR DESCRIPTION
**Pull request by vlandham**
This is the start of a 'Categorical Scatterplot' or catscat for short. 

It provides a very basic starting point for building a scatterplot that:
* has an x-axis as a categorical variable
* can use a metric to define the point shape
* can draw additional threshold lines on it.

screen shot 2017-11-14 at 10 55 59 am

The image above is the use of the catscat chart on Superset provided dummy data. 

An incomplete list of Open questions we need to figure out before this is ready to go. Most are a trade-off between time and capabilities:

* What side controls are needed for appropriate grouping? Should we try to reuse the existing controls from the bubble chart (which i find confusing) or build out our own?
* How should color-by work? does it need to be added to the `groupby` in the query, or can it be based on a metric or column value?
* Do we need "sort by" capabilities? Should this be part of the query - or done in the visualization?
* what customization controls are needed for configuring the x-axis? Should we provide support for tick customization? 
* what customization controls are needed for configuring the y-axis?
* what customization controls are needed for configuring the thresholds? Do we need to allow inputing color? thickness? other attributes?
* what level of interaction is required. Do we need mouse-over tooltips on dots?

### CODE REVIEW

**Comment by zymergen-jchen** [comment]
_What side controls are needed for appropriate grouping? Should we try to reuse the existing controls from the bubble chart (which i find confusing) or build out our own?_
Might be different with bubble chart. Not sure.

_How should color-by work? does it need to be added to the groupby in the query, or can it be based on a metric or column value?_
Color-by should be added to the groupby in the query.

_Do we need "sort by" capabilities? Should this be part of the query - or done in the visualization?
what customization controls are needed for configuring the x-axis? Should we provide support for tick customization?_
We need 'sort by'. User should be able to define it on side control. In our case, user probably want to sort by something like experiment_complete_time.

_what customization controls are needed for configuring the y-axis?_
At least, y-axis label, and y-axis range.

_what customization controls are needed for configuring the thresholds? Do we need to allow inputing color? thickness? other attributes?_
Nice to have, not essential.

_what level of interaction is required. Do we need mouse-over tooltips on dots?_
Yes. We at least need mouse-over tooltips, and un-select by clicking on legend.

File: superset/assets/visualizations/catscat.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+  // api: https://github.com/d3/d3-3.x-api-reference/blob/master/Quantitative-Scales.md#linear
+  const yMin = d3.min(data, d => d3.min(d.values, v => v.y));
+  const yMax = d3.max(data, d => d3.max(d.values, v => v.y));
+  const yScale = d3.scale.linear()
+    .domain([yMin, yMax])
+    .range([plotHeight, 0]);
+
+  // as this is a categorical scatterplot, our x scale will be ordinal. 
+  // api: https://github.com/d3/d3-3.x-api-reference/blob/master/Ordinal-Scales.md#ordinal_rangePoints
+  const xScale = d3.scale.ordinal()
```
Comment (zymergen-jchen) [inline superset/assets/visualizations/catscat.js]: We might want xScale to be linear, too. The categories will be shown by colorScale and shapeScale.

File: superset/assets/visualizations/catscat.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    .domain(d3.range(data.length))
+    .rangePoints([0, plotWidth]);
+
+  const shapeValuesAll = data.map(d => d.values.map(v => v.shape));
+  const shapeValues = d3.set([].concat(...shapeValuesAll)).values();
```
Comment (zymergen-jchen) [inline superset/assets/visualizations/catscat.js]: How do I read this line?

File: superset/assets/visualizations/catscat.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+import './catscat.css';
+import '../stylesheets/d3tip.css';
+
+/**
+ * computedProps
+ * returns derived properties and functions based on input.
+ * Used to conceptually distinguish the 'setting up' of
+ * a chart from the actual 'dom manipulation' to create the chart.
+ * @param {Object} props : Input properties to use to compute
+ * @returns {Object} : Object of computed props.
```
Comment (zymergen-jchen) [inline superset/assets/visualizations/catscat.js:146]: ?

File: superset/assets/visualizations/catscat.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+ * Main Visualization method.
+ * Takes the slice and json objects and
+ * modifies DOM to create chart.
+ * 
+ * @param {Object} slice 
+ * @param {Object} json 
+ */
+function scatCatViz(slice, json) {
+  // this removes everything that might already
+  // be in the vis's container.
```
Comment (zymergen-jchen) [inline superset/assets/visualizations/catscat.js]: What is vis's contrainer?

File: superset/assets/visualizations/catscat.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    .attr('class', 'd3-tip')
+    .direction('n')
+    .offset([-5, 0])
+    .html(renderTooltip);
+  svg.call(tip);
+
+
+  // data binding and enter selection
+  // for outer data array
+  const band = g.selectAll('.band')
```
Comment (zymergen-jchen) [inline superset/assets/visualizations/catscat.js:350]: ?

File: superset/assets/visualizations/catscat.js (hunk 2 above)
Comment (vlandham) [reply superset/assets/visualizations/catscat.js]: flattens the array of arrays.

the dot dot dot is the spread operator: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator

**Comment by vlandham** [comment]
Now sorts by the 'colorby' value. 
Legend is below the graph. Currently a crude guess at the height of the legend is made. This might need to be improved as we test with more data. 
Legend is not interactive, but that seems like a good first start. 

screen shot 2017-12-14 at 9 22 10 am

**Pull request closed without merging**